athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/ElasticsearchMetadataHandler.java

```diff
@@ -0,0 +1,282 @@
+ * <p>
+ * For more examples, please see the other connectors in this repository (e.g. athena-cloudwatch, athena-docdb, etc...)
+ */
+public class ElasticsearchMetadataHandler
+        extends GlueMetadataHandler
+{
+    private static final Logger logger = LoggerFactory.getLogger(ElasticsearchMetadataHandler.class);
+
+    // Used to denote the 'type' of this connector for diagnostic purposes.
+    private static final String SOURCE_TYPE = "elasticsearch";
+
+    //The Env variable name used to indicate that we want to disable the use of Glue DataCatalog for supplemental
+    //metadata and instead rely solely on the connector's schema inference capabilities.
+    private static final String GLUE_ENV = "disable_glue";
+
+    private final AWSGlue awsGlue;
+    private final AwsRestHighLevelClientFactory clientFactory;
+
+    public ElasticsearchMetadataHandler()
+    {
+        //Disable Glue if the env var is present and not explicitly set to "false"
+        super((System.getenv(GLUE_ENV) != null && !"false".equalsIgnoreCase(System.getenv(GLUE_ENV))), SOURCE_TYPE);
+        this.awsGlue = getAwsGlue();
+        this.clientFactory = ElasticsearchHelper.getClientFactory();
+
+        ElasticsearchHelper.setDomainMapping();
+    }
+
+    @VisibleForTesting
+    protected ElasticsearchMetadataHandler(AWSGlue awsGlue,
+                                           EncryptionKeyFactory keyFactory,
+                                           AWSSecretsManager awsSecretsManager,
+                                           AmazonAthena athena,
+                                           String spillBucket,
+                                           String spillPrefix,
+                                           AwsRestHighLevelClientFactory clientFactory)
+    {
+        super(awsGlue, keyFactory, awsSecretsManager, athena, SOURCE_TYPE, spillBucket, spillPrefix);
+        this.awsGlue = awsGlue;
+        this.clientFactory = clientFactory;
+
+        ElasticsearchHelper.setDomainMapping();
+    }
+
+    /**
+     * Used to get the list of domains (aka databases) for the Elasticsearch service.
+     * @param allocator Tool for creating and managing Apache Arrow Blocks.
+     * @param request Provides details on who made the request and which Athena catalog they are querying.
+     * @return A ListSchemasResponse which primarily contains a Set<String> of schema names and a catalog name
+     * corresponding the Athena catalog that was queried.
+     */
+    @Override
+    public ListSchemasResponse doListSchemaNames(BlockAllocator allocator, ListSchemasRequest request)
+    {
+        logger.info("doListSchemaNames: enter - " + request);
+
+        if (ElasticsearchHelper.isAutoDiscoverEndpoint()) {
+            // Since adding/deleting domains in Amazon Elasticsearch Service doesn't re-initialize
+            // the connector, the domainMap needs to be refreshed each time for Amazon ES.
+            ElasticsearchHelper.setDomainMapping();
+        }
```

Comment:
Is there a reason we can't check this in the constructor and configure ourselves accordingly instead of checking this on every call? This also has the nicety of not needing / using the 'static' configuration construct which can lead or other maintainability challenges.

Reply:
There is a method behind the madness, which is not to say that it doesn't come at a cost... Implementing this in the constructor means that the domain mapping is only set once. The implication for Amazon Elasticsearch Service users is that each time they delete or add a domain in their Elasticsearch service, they will need to refresh/restart the connector. This makes it less than optimal for them. The other implication is, that both Metadeta handler and Record handler will each have to have their own copy of the data, as opposed to having it in one central place (duplication = more chances for something to break). Of course, it comes at a performance cost (only for Amazon ES users), so I'm on the fence about it. If you prefer the constructor approach, I will add language to the README to state the aforementioned dependency. Otherwise, I will keep it as-is. Let me know which approach you prefer.

Reply:
Gone due to refactoring.